### PR TITLE
Multiapplication Support

### PR DESCRIPTION
**Pull request by brianzhang2101**
Generalising the CMS to enable other teams to utilise it. Improving user experience & security issues with sharing files in one large repository.

* New frontend table maps from URL, Name to an isolated _root_ in the filesystem table which contains all files/directories branches
    * Helper function added to set these up
* Schema now supports the concept of groups and permissions to read, write and delete a collection of _applications_ or _frontends_ (helper functions not in place yet)
* FrontendID now uses UUID
* Updated and reorganised Postgres testing data to better emulate data flow better

Update (rest of the specifics above are same):

* Small refactor on schema to now have an **ID** as primary key on frontend table and a separate **root** column representing the root inside the filesystem table
* Setup functions will handle UUID generation (no more code-side generation, which was inconsistent and confusing to read in tests)

### CODE REVIEW

File: backend/database/repositories/frontends.go [hunk 1]
```diff
@@ -1,15 +1,32 @@
...
 type frontendsRepository struct {
 	embeddedContext
 }
 
 const InvalidFrontend = -1
 
+func NewFrontendRepo(frontEndID uuid.UUID, logicalName string, URL string, embeddedContext embeddedContext) (filesystemRepository, error) {
+	err := embeddedContext.ctx.Query("SELECT new_frontend($1, $2, $3)", []interface{}{frontEndID, logicalName, URL}, &frontEndID)
+	if err != nil {
+		return filesystemRepository{}, err
```
Comment (Varun-Sethu) [inline backend/database/repositories/frontends.go:16]: I feel like moving on there should be better handling of errors, perhaps we should wrap this error up in a secondary error to decorate it with a bit more context

File: backend/endpoints/dependency_factory.go [hunk 2]
```diff
@@ -25,14 +26,16 @@ type (
...
+		FrontEndID  uuid.UUID
+		LogicalName string
+		URL         string
 	}
 )
 
 // GetFilesystemRepo is the constructor for FS repos
-func (dp DependencyProvider) GetFilesystemRepo() repos.FilesystemRepository {
-	return repos.NewFilesystemRepo(contexts.GetDatabaseContext())
+func (dp DependencyProvider) GetFilesystemRepo() (repos.FilesystemRepository, error) {
```
Comment (Varun-Sethu) [inline backend/endpoints/dependency_factory.go:37]: IMO the dependency provider should be decorating the error returned by the repository factory

File: backend/endpoints/filesystem_endpoints.go [hunk 3]
```diff
@@ -11,7 +11,14 @@ import (
 // Defines endpoints consumable via the API
 func GetEntityInfo(form ValidInfoRequest, df DependencyFactory) handlerResponse[EntityInfoResponse] {
 	log := df.GetLogger()
-	fsRepo := df.GetFilesystemRepo()
+	fsRepo, err := df.GetFilesystemRepo()
+
+	if err != nil {
+		return handlerResponse[EntityInfoResponse]{
+			Status:   http.StatusNotFound,
```
Comment (Varun-Sethu) [inline backend/endpoints/filesystem_endpoints.go:18]: I feel like this is an inappropriate response for failing to get the FS repo

File: backend/endpoints/filesystem_endpoints.go [hunk 4]
```diff
@@ -33,7 +40,13 @@ func GetEntityInfo(form ValidInfoRequest, df DependencyFactory) handlerResponse[
 // CreateNewEntity is the public handler for constructing and creating new entities
 func CreateNewEntity(form ValidEntityCreationRequest, df DependencyFactory) handlerResponse[NewEntityResponse] {
 	log := df.GetLogger()
-	fsRepo := df.GetFilesystemRepo()
+	fsRepo, err := df.GetFilesystemRepo()
+	if err != nil {
+		return handlerResponse[NewEntityResponse]{
+			Status:   http.StatusNotFound,
```
Comment (Varun-Sethu) [inline backend/endpoints/filesystem_endpoints.go:46]: Same here ^^

**Review by Varun-Sethu — CHANGES_REQUESTED** [review]
Looks good, some minor comments though and also:
 - Since I think the original goal of the ticket kinda changed its probs worth adding a full list of changes in the PR description
 - I think you may need to regenerate mocks as you did end up refactoring the repository layer quite a bit.
     - You will need to update https://github.com/csesoc/website/blob/3c8d63f5e6f408953dadf2961c7733744692d07b/backend/database/repositories/repository_interfaces.go#L3  to be `//go:generate mockgen -source=repository_interfaces.go -destination=mocks/repositories_mock.go -package=mocks`
     - And then run `go generate ./...` at the top level in the backend directory

File: postgres/up/02-create_frontend_table.sql [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+/* Maps frontend URL to its name and root within filesystem */ 
 DROP TABLE IF EXISTS frontend;
-CREATE TABLE frontend (
-  FrontendID  SERIAL PRIMARY KEY,
-  FrontendURL VARCHAR(100)
+CREATE TABLE frontend 
+(
+  ID  uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
```
Comment (Varun-Sethu) [inline postgres/up/02-create_frontend_table.sql:7]: referencing discord discussion, so the FE ID is declared here

File: postgres/up/05-create_filesystem_table.sql [hunk 6]
```diff
@@ -61,6 +62,24 @@ BEGIN
...
+-- TODO: Add a delete frontend function
+DROP FUNCTION IF EXISTS new_frontend;
+CREATE OR REPLACE FUNCTION new_frontend (frontendIDP uuid, logicalNameP VARCHAR, URLP VARCHAR) RETURNS uuid
+LANGUAGE plpgsql
+AS $$
+DECLARE
+  frontendID frontend.ID%type;
+BEGIN
+  INSERT INTO frontend VALUES (frontendIDP, logicalNameP, URLP)
+    RETURNING ID INTO frontendID;
```
Comment (Varun-Sethu) [inline postgres/up/05-create_filesystem_table.sql:74]: then you create a new frontend here and take the frontend ID UUID

File: postgres/up/05-create_filesystem_table.sql [hunk 7]
```diff
@@ -61,6 +62,24 @@ BEGIN
...
+       
+  -- TODO: Temporarily setting the OwnedBy field. Going to be deprecated soon.
+  INSERT INTO filesystem (EntityID, LogicalName, IsDocument, Parent, OwnedBy)
```
Comment (Varun-Sethu) [inline postgres/up/05-create_filesystem_table.sql:77]: and then create a new file tree where the ID of the root is the frontendID, basically you overload the value you generate when creating a frontend to act as both the ID for the FE table and the ID for the root entity for that table

Comment (Varun-Sethu) [reply postgres/up/05-create_filesystem_table.sql:77]: IMO u shouldn't do this, you should maintain a separate column within the FE table that references that root, additionally we should have a foreign key constraint somewhere that enforces that

**Review by Varun-Sethu — CHANGES_REQUESTED** [review]

File: backend/database/repositories/tests/filesystem_test.go [hunk 8]
```diff
@@ -15,8 +15,11 @@ import (
 )
 
 var (
-	repo        = repositories.NewFilesystemRepo(contexts.GetDatabaseContext())
-	testContext = repo.GetContext().(*contexts.TestingContext)
+	frontendID          = uuid.New()
+	frontendLogicalName = "CSESoc Test"
+	frontendURL         = "http://localhost:3001"
+	testContext         = contexts.GetDatabaseContext().(*contexts.TestingContext)
+	repo, _             repositories.FilesystemRepository
```
Comment (Varun-Sethu) [inline backend/database/repositories/tests/filesystem_test.go:22]: no :(

**Review by Varun-Sethu — APPROVED** [review]

**Comment by ShunyaoLiang** [comment]
Thanks for the meeting Brian. Looks great to me, just want there to be unit tests for multiple frontend repositories at once.

**Comment by brianzhang2101** [comment]
Luckily this test did actually spot a bug where `GetIDWithPath()` forgets to query specifically within the FE's filetree, thanks @ShunyaoLiang

**Review by ShunyaoLiang — APPROVED** [review]
That new test looks good to me. Great work Brian!

**Pull request merged**